### ISSUE TITLE
predict() is O(n_estimators) per call with no caching; document or snapshot

### ISSUE THREAD

**Issue by sachncs** [sachncs/gradient-regularized-newton-boosting-trees#37]
## Problem

`grnbt/boosting.py:243-251` (`BaseBoosting.predict`) and `grnbt/boosting.py:574-582` (`MultiClassNewtonBoosting.predict`) iterate over the full ensemble (`self.trees`) and call `tree.predict(x)` for every tree on every `predict` call. There is no caching of intermediate `F_k` predictions.

For `n_estimators=100` and `x.shape[0]=10000`, the cost is `100 × predict_cost` per call. For `n_estimators=1000` it is `1000 × predict_cost`.

The README's Quick Start uses `n_estimators=100` by default, and `docs/experiments.md:144-146` says "Wine experiment: ~10 seconds on a modern CPU." — meaning at the small-data scale the cost is bearable. But scaling up to larger ensembles on larger inputs reveals the linear cost.

## Evidence

- `grnbt/boosting.py:253-255` — `for tree in self.trees: f_out += self.learning_rate * tree.predict(x)`.
- `grnbt/boosting.py:577-578` — same pattern in multi-class.
- No `_cache` or `_intermediate_F` attribute on `BaseBoosting`.
- `docs/architecture.md:155-158` acknowledges "Performance-critical inner loops (e.g., tree traversal) use Python `for` loops; for large-scale production use the histogram extension or a compiled backend would be appropriate."

## Root Cause

The training loop materializes trees but does not materialize predictions. `predict` replays the full tree sequence on every call.

## Impact

- **Performance**: linear in `n_estimators` per `predict` call.
- **Scaling ceiling**: large ensembles are unusable for online inference.
- **No early-stopping benefit**: even if a user implements early stopping in `fit`, the final `predict` still replays all stored trees.

## Recommendation

Three options:

**Option A — snapshot intermediate `F_k`** at periodic intervals (e.g., every 10 trees). Predict then only needs to replay from the most recent snapshot. This is the standard "prediction cache" pattern.

**Option B — document the O(n_estimators) cost explicitly** in the `predict` docstring so users know what to expect.

**Option C — provide a `predict_oof` (out-of-fold) helper for cross-validation use cases**, which often needs `predict(x)` to be O(1) per fold.

Option A is the most useful but adds memory overhead. Option B is the cheapest. The repository is research-grade, so Option B may be acceptable as the default with Option A as a future enhancement.

## Acceptance Criteria

- [ ] `predict` docstring explicitly states O(n_estimators) cost.
- [ ] (Optional) A `predict_cached(every_k=10)` method exists that snaps intermediate F_k predictions at intervals of `every_k`.

## Verification

```bash
python -c "
import time, numpy as np
from grnbt import VanillaNewtonBoosting, MSELoss
X = np.random.randn(10000, 8)
y = X[:, 0] + np.random.randn(10000)
m = VanillaNewtonBoosting(loss=MSELoss(), n_estimators=100, max_depth=3)
m.fit(X[:5000], y[:5000])
t0 = time.time(); m.predict(X[5000:]); print(f'predict: {time.time() - t0:.3f}s')
"
```

Document the cost per ensemble size.

## Dependencies

- Issue #19 (template-method refactor of `MultiClassNewtonBoosting.fit`) may interact with where snapshots are stored.

## Risk

Low. Option A adds memory; Option C adds API surface. None change correctness.

**Comment by sachncs** [comment]
Resolved in PR #40.
